### PR TITLE
Support for default table (and cells) properties

### LINKED ISSUE #8502: No way to set default alignment for images or tables.
When you insert images / tables there's no way to pick the default alignment. 

I looked at the alignment for images and there is a 'styles' which has an isDefault property but that's not actually working.

I also can't find any documentation on how to do this with tables either.

### LINKED ISSUE #9219: Default table properties via configuration
## 📝 Provide a description of the new feature

_What is the expected behavior of the proposed feature?_

e.g. How do we configure the default table styles? I want a 1px border with certain colors to be pre-filled.

Example:

```js
{
  table: {
    tableProperties: {
      defaults: {
        borderColor: 'red',
        borderWidth: '1',
        borderStyle: 'solid'
      }
   }
}
```

Upon the user inserting a table, they should see:

image

History: To avoid confusion, this remakes #6363's [original post](https://github.com/ckeditor/ckeditor5/issues/6363#issue-573975234) as its topic changed and it was closed. The new topic at #6841 is speaking about something different than setting default (clearing the value vs setting `none` to negate?)

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/contributing/git-commit-message-convention.html))

Feature (table): The [`TablePropertiesEditing`](https://ckeditor.com/docs/ckeditor5/latest/api/module_table_tableproperties_tablepropertiesediting-TablePropertiesEditing.html) plugin allows specifying default properties for the new inserted tables. Closes #8502. Closes #9219.

Feature (table): The [`TableCellPropertiesEditing`](https://ckeditor.com/docs/ckeditor5/latest/api/module_table_tablecellproperties_tablecellpropertiesediting-TableCellPropertiesEditing.html) plugin allows specifying default properties of all cells when inserting the new table.

Other (table): The `TableUtils#createTable()` is now decorated and fires an event when executing the function. Read more about [decorating object methods](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/deep-dive/observables.html#decorating-object-methods).

Other (table): The `TableUtils#insertRows()` and `TableUtils#insertColumns()` returns an array that contains [the created model elements](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_model_element-Element.html).

### CODE REVIEW

**Comment by Mgsy** [comment]
PR is ready for testing @ckeditor/qa-team.

**Comment by FilipTokarski** [comment]
As it comes to default table alignment, I guess those issues will become more frequent now:
- https://github.com/ckeditor/ckeditor5/issues/7749
- https://github.com/ckeditor/ckeditor5/issues/8412

**Comment by FilipTokarski** [comment]
## Steps to reproduce:
1. Align table left
2. Add block quote after the table

## Result
No space between table border and block quote border:
  
0_table2

**Comment by FilipTokarski** [comment]
## Steps to reproduce
1. Align table left or right
2. Add html embed after the table

## Result
Html embed is overlapping table border:

https://user-images.githubusercontent.com/34380544/113144528-a12a0000-922d-11eb-92e6-1b6c763f2178.mp4

**Comment by FilipTokarski** [comment]
Looks ok 👍

File: packages/ckeditor5-table/src/tablecellproperties.js [hunk 1]
```diff
@@ -64,6 +64,31 @@ export default class TableCellProperties extends Plugin {
...
+ *
+ *		const tableConfig = {
+ *			tableCellProperties: {
+ *				defaultProperties: {
+ *				    borderStyle: 'dotted',
+ *				    borderColor: 'hsl(120, 75%, 60%)',
+ *				    borderWidth: '2px',
+ *				    horizontalAlignment: 'right',
+ *				    verticalAlignment: 'bottom'
+ *				}
```
Comment (pomek) [inline packages/ckeditor5-table/src/tablecellproperties.js:77]:
```suggestion
 *				defaultProperties: {
 *					borderStyle: 'dotted',
 *					borderColor: 'hsl(120, 75%, 60%)',
 *					borderWidth: '2px',
 *					horizontalAlignment: 'right',
 *					verticalAlignment: 'bottom'
 *				}
```

File: packages/ckeditor5-table/src/tableproperties.js [hunk 2]
```diff
@@ -65,6 +65,29 @@ export default class TableProperties extends Plugin {
...
+ * * The default styles for the new created table (`tableProperties.defaultProperties`):
+ *
+ *		const tableConfig = {
+ *			tableProperties: {
+ *				defaultProperties: {
+ *				    borderStyle: 'dashed',
+ *				    borderColor: 'hsl(0, 0%, 90%)',
+ *				    borderWidth: '3px',
+ *				    alignment: 'left'
+ *				}
```
Comment (pomek) [inline packages/ckeditor5-table/src/tableproperties.js:77]:
```suggestion
 *				defaultProperties: {
 *					borderStyle: 'dashed',
 *					borderColor: 'hsl(0, 0%, 90%)',
 *					borderWidth: '3px',
 *					alignment: 'left'
 *				}
```

File: packages/ckeditor5-table/src/tableutils.js [hunk 3]
```diff
@@ -94,15 +94,22 @@ export default class TableUtils extends Plugin {
 	 * @param {Number} [options.columns=2] The number of columns to create.
 	 * @param {Number} [options.headingRows=0] The number of heading rows.
 	 * @param {Number} [options.headingColumns=0] The number of heading columns.
+	 * @param {String} [options.defaultProperties={}] The default properties for the created table.
```
Comment (jacekbogdanski) [inline packages/ckeditor5-table/src/tableutils.js:97]:
```suggestion
	 * @param {Object} [options.defaultProperties={}] The default properties for the created table.
```

Comment (jacekbogdanski) [reply packages/ckeditor5-table/src/tableutils.js:97]: Shouldn't be it named just `attributes` or `tableAttributes` (to distinguish it from cell attributes)? Looking at the code this option is specific to the default properties feature.

File: packages/ckeditor5-table/src/tableutils.js [hunk 4]
```diff
@@ -94,15 +94,22 @@ export default class TableUtils extends Plugin {
...
+	 * @param {Object} [options.defaultCellProperties={}] The default cell properties in the created table.
```
Comment (jacekbogdanski) [inline packages/ckeditor5-table/src/tableutils.js:98]: Also, shouldn't be just `cellAttributes`?

File: packages/ckeditor5-table/src/tableutils.js [hunk 5]
```diff
@@ -748,13 +769,15 @@ export default class TableUtils extends Plugin {
 // @param {Number} insertAt The row index of row insertion.
 // @param {Number} rows The number of rows to create.
 // @param {Number} tableCellToInsert The number of cells to insert in each row.
-function createEmptyRows( writer, table, insertAt, rows, tableCellToInsert, attributes = {} ) {
+// @param {Object} options
+// @param {Object} options.cellProperties The cell attributes.
+function createEmptyRows( writer, table, insertAt, rows, tableCellToInsert, options ) {
```
Comment (jacekbogdanski) [inline packages/ckeditor5-table/src/tableutils.js:774]: Not sure if we gain something from making `attributes` argument more general. From what I see `createEmptyRows` is not used with anything other than just model element attributes. At the end of the passing chain we are using `createEmptyTableCell` method accepting plain attributes:
https://github.com/ckeditor/ckeditor5/blob/912cb064cd7aaad47a6d193b6b74de4193414ad5/packages/ckeditor5-table/src/utils/common.js#L35-L42

Maybe we could make it simple for now?

File: packages/ckeditor5-table/docs/features/table.md [hunk 6]
```diff
@@ -205,6 +205,45 @@ ClassicEditor
 	.catch( ... );
 ```
 
+### Default table and cell styles
+
+You can specify the default styles for the table and cells that will be applied while creating the new table by passing the `defaultProperties` object to {@link module:table/table~TableConfig#tableProperties} for tables, and {@link module:table/table~TableConfig#tableCellProperties} for cells.
```
Comment (jacekbogdanski) [inline packages/ckeditor5-table/docs/features/table.md:210]:
```suggestion
You can specify the default styles for the table and cells that will be applied while creating a new table. To configure default table styles, pass the `defaultProperties` object to {@link module:table/table~TableConfig#tableProperties} configuration option. Similarly, cell styles can be configured by setting `defaultProperties` on {@link module:table/table~TableConfig#tableCellProperties} option.
```

File: packages/ckeditor5-table/tests/_utils/utils.js [hunk 7]
```diff
@@ -32,19 +32,26 @@ const WIDGET_TABLE_CELL_CLASS = 'ck-editor__editable ck-editor__nested-editable'
  *		};
  *
  * @param {Array.<Array.<String>|Object>} tableData
- * @param {Object} [attributes] Optional table attributes: `headingRows` and `headingColumns`.
+ * @param {Object} [attributes={}] Optional table attributes.
+ * @param {Number} attributes.headingRows Number of heading rows.
+ * @param {Number} attributes.headingColumns Number of heading columns.
+ * @param {Object} attributes.defaultProperties Default styles for the table.
+ * @param {Object} attributes.defaultCellProperties Default styles for all cells.
```
Comment (jacekbogdanski) [inline packages/ckeditor5-table/tests/_utils/utils.js:39]: It's probably more like `options` than `attributes` in that case.

**Review by jacekbogdanski — CHANGES_REQUESTED** [review]
I'm getting some error when merging cells with default properties:

https://user-images.githubusercontent.com/10279586/113712693-ba77f400-96e6-11eb-8cd6-e7c96a8e7dc2.mov

File: packages/ckeditor5-table/src/commands/insertcolumncommand.js [hunk 8]
```diff
@@ -78,6 +78,15 @@ export default class InsertColumnCommand extends Command {
...
+		const insertColumnsOptions = {
+			columns: 1,
+			at: insertBefore ? column : column + 1
+		};
+
+		if ( editor.plugins.has( 'TableCellPropertiesEditing' ) ) {
+			insertColumnsOptions.defaultCellProperties = editor.config.get( 'table.tableCellProperties.defaultProperties' );
+		}
+
+		tableUtils.insertColumns( table, insertColumnsOptions );
```
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/commands/insertcolumncommand.js:90]: Regarding the `InsertTableCommand` – typing `editor.execute('insertTable')` in the console should work the same way as inserting the table by UI. So, extending the constructor arguments seems to be the only correct solution here.

The proposed solution, specifying default options to constructors looks good but will not work if a user updates the runtime editor configuration.

**Comment by pomek** [comment]
Before introducing this change: 3acb05d419c965eb22dbbab63c274b4920b481b6, everything worked.

```js
editor.setData(`<p style="text-align: justify;"><p style="text-align: center;">JavaScript</p></p>`)

editor.getData()
"<p style="text-align:justify;">&nbsp;</p><p style="text-align:center;">JavaScript</p><p>&nbsp;</p>"
```

File: packages/ckeditor5-table/src/commands/inserttablecommand.js [hunk 9]
```diff
@@ -47,14 +47,25 @@ export default class InsertTableCommand extends Command {
...
+
+		const createTableOptions = Object.assign( {}, options );
+
+		if ( editor.plugins.has( 'TablePropertiesEditing' ) ) {
+			createTableOptions.defaultProperties = editor.config.get( 'table.tableProperties.defaultProperties' );
+		}
+
+		if ( editor.plugins.has( 'TableCellPropertiesEditing' ) ) {
+			createTableOptions.defaultCellProperties = editor.config.get( 'table.tableCellProperties.defaultProperties' );
+		}
```
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/commands/inserttablecommand.js:63]: Two points against this solution:

- functions that do not do too much, just assign single property to the object,
- side effect (modifying arguments) - utils do not return anything, just update received arguments.

File: packages/ckeditor5-table/src/tablecellproperties.js [hunk 10]
```diff
@@ -64,6 +64,31 @@ export default class TableCellProperties extends Plugin {
...
+ * * The default styles for the cell while creating a new table (`tableCellProperties.defaultProperties`):
+ *
+ *		const tableConfig = {
+ *			tableCellProperties: {
+ *				defaultProperties: {
+ *					borderStyle: 'dotted',
+ *					borderColor: 'hsl(120, 75%, 60%)',
+ *					borderWidth: '2px',
+ *					horizontalAlignment: 'right',
+ *					verticalAlignment: 'bottom'
```
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/tablecellproperties.js:76]: I'll add those to API docs.

File: packages/ckeditor5-table/src/tableproperties.js [hunk 11]
```diff
@@ -65,6 +65,29 @@ export default class TableProperties extends Plugin {
...
+ *		const tableConfig = {
+ *			tableProperties: {
+ *				defaultProperties: {
+ *					borderStyle: 'dashed',
+ *					borderColor: 'hsl(0, 0%, 90%)',
+ *					borderWidth: '3px',
+ *					alignment: 'left'
+ *				}
+ *			}
+ *		}
```
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/tableproperties.js:79]: Fixed in 5ad7484.

File: packages/ckeditor5-table/src/tablecellproperties.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (pomek) [reply packages/ckeditor5-table/src/tablecellproperties.js:76]: Fixed in 5ad7484.

File: packages/ckeditor5-table/src/tableutils.js (hunk 3 above)
Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js:97]: +1 for `tableAttributes`. The word "default" means nothing in Table Utils' context.

Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js:97]: I'd go with the `tableProperties` name to align the option to the plugin. Same for cell properties.

Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js:97]: But it's hard to be consistent :/

Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js]: Fixed in c7b23f2.

File: packages/ckeditor5-table/src/tableutils.js (hunk 4 above)
Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js]: Fixed in c7b23f2.

File: packages/ckeditor5-table/src/tableutils.js (hunk 5 above)
Comment (pomek) [reply packages/ckeditor5-table/src/tableutils.js]: Done in c7b23f2.

File: packages/ckeditor5-table/tests/_utils/utils.js (hunk 7 above)
Comment (pomek) [reply packages/ckeditor5-table/tests/_utils/utils.js]: Updated in https://github.com/ckeditor/ckeditor5/pull/9393/commits/88046d8cf94998dd3bc40c7ffc6317c2676fc369.

**Comment by pomek** [comment]
Today we discussed another approach to have default table/cell properties. Instead of modifying the current commands, let's try to decorate them and apply the properties.

**Comment by pomek** [comment]
Problem to solve: single undo step after creating a table with applied styles. As for now, we have three steps: 

1. created table
2. applied table properties
3. applied cell properties

**Comment by pomek** [comment]
> I'm getting some error when merging cells with default properties:

After redesigning the feature to decorate commands/plugins, the error does not occur anymore.

**Comment by pomek** [comment]
The PR is ready for review once again. 

Now, the proposed solution uses decorated methods from the `TableUtils` plugin. Thanks to that, table commands do not have to know about table/cell properties. The `TableProperties` and `TableCellProperties` listen to proper events and modify the table before inserting it into the document.

Splitting a column does not inherit "the parent" properties but it should be resolved in another issue – https://github.com/ckeditor/ckeditor5/issues/7246.

cc: @niegowski

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module table/converters/table-layout-post-fixer
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:7]:
```suggestion
 * @module table/converters/table-cell-default-properties-post-fixer
```

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	let wasFixed = false;
+
+	for ( const entry of changes ) {
+		if ( entry.type != 'insert' ) {
+			continue;
+		}
+
+		// Fix table on adding/removing table cells and rows.
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:54]:
```suggestion
		// Fix table cells on adding table rows.
```

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+				if ( shouldApplyDefaultCellProperties( tableCell ) ) {
+					// ...and if the cell has no properties, apply the default.
+					writer.setAttributes( cellProperties, tableCell );
+
+					wasFixed = true;
+				}
+			}
+		}
+
+		// Fix table cell on adding/removing table cells and rows.
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:70]:
```suggestion
		// Fix table cell after adding it.
```

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+		if ( entry.name == 'tableCell' ) {
+			const tableCell = entry.position.nodeAfter;
+
+			if ( shouldApplyDefaultCellProperties( tableCell ) ) {
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:74]:
```suggestion
			// Check cell's properties...
			if ( shouldApplyDefaultCellProperties( tableCell ) ) {
```

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+}
+
+// Checks whether the default properties should be applied for the specified cell.
+//
+// The default properties will be applied only if the cell does not contain any "visual" properties.
+//
+// @param {module:engine/model/element~Element} tableCell
+// @returns {Boolean}
+function shouldApplyDefaultCellProperties( tableCell ) {
+	const attrs = [ ...tableCell.getAttributeKeys() ];
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:93]:
```suggestion
	const attrs = Array.from( tableCell.getAttributeKeys() );
```

To be more explicit.

File: packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+
+	return attrs.some( attributeName => TABLE_CELL_PROPERTIES.includes( attributeName ) ) === false;
```
Comment (niegowski) [inline packages/ckeditor5-table/src/converters/table-cell-default-properties-post-fixer.js:95]: Maybe this line would be easier to read if we flip this helper logic to `hasCellProperties` ?

File: packages/ckeditor5-table/tests/_utils/utils.js [hunk 18]
```diff
@@ -32,20 +32,26 @@ const WIDGET_TABLE_CELL_CLASS = 'ck-editor__editable ck-editor__nested-editable'
...
-		enforceWrapping: true
+		enforceWrapping: true,
+		cellProperties: options.cellProperties
 	} );
 
-	return `<table${ formatAttributes( attributes ) }>${ tableRows }</table>`;
+	delete options.cellProperties;
+
+	return `<table${ formatAttributes( options ) }>${ tableRows }</table>`;
 }
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/_utils/utils.js:55]: I don't like the idea of mixing the table element attributes with the `cellProperties` meta field. Maybe we could leave the `attributes` as it was and introduce additional `options` parameter?

File: packages/ckeditor5-table/tests/commands/insertcolumncommand.js [hunk 19]
```diff
@@ -14,14 +14,15 @@ import TableEditing from '../../src/tableediting';
 import { assertSelectedCells, modelTable } from '../_utils/utils';
 
 import InsertColumnCommand from '../../src/commands/insertcolumncommand';
+import { UndoEditing } from '@ckeditor/ckeditor5-undo';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/commands/insertcolumncommand.js:17]: Please fix the order of imports.

File: packages/ckeditor5-table/tests/commands/insertrowcommand.js [hunk 20]
```diff
@@ -14,14 +14,15 @@ import TableSelection from '../../src/tableselection';
 import { assertSelectedCells, modelTable } from '../_utils/utils';
 
 import InsertRowCommand from '../../src/commands/insertrowcommand';
+import { UndoEditing } from '@ckeditor/ckeditor5-undo';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/commands/insertrowcommand.js:17]: Please fix the order of imports.

File: packages/ckeditor5-table/tests/commands/insertrowcommand.js [hunk 21]
```diff
@@ -308,6 +309,44 @@ describe( 'InsertRowCommand', () => {
 					[ '10', '11', '12' ]
 				] ) );
 			} );
+
+			it( 'should create a single undo step when inserting a column', () => {
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/commands/insertrowcommand.js:313]:
```suggestion
			it( 'should create a single undo step when inserting a row', () => {
```

File: packages/ckeditor5-table/tests/converters/table-cell-default-properties-post-fixer.js [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+ */
+
+import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
+import VirtualTestEditor from '@ckeditor/ckeditor5-core/tests/_utils/virtualtesteditor';
+import { getData as getModelData, parse, setData as setModelData } from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
+
+import TableEditing from '../../src/tableediting';
+import { modelTable } from './../_utils/utils';
+import UndoEditing from '@ckeditor/ckeditor5-undo/src/undoediting';
+import { assertEqualMarkup } from '@ckeditor/ckeditor5-utils/tests/_utils/utils';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/converters/table-cell-default-properties-post-fixer.js:13]: Please fix the order of imports.

File: packages/ckeditor5-table/tests/tablecellproperties/tablecellpropertiesediting.js [hunk 23]
```diff
@@ -19,9 +19,11 @@ import TableCellVerticalAlignmentCommand from '../../src/tablecellproperties/com
 import TableCellPaddingCommand from '../../src/tablecellproperties/commands/tablecellpaddingcommand';
 import TableCellBackgroundColorCommand from '../../src/tablecellproperties/commands/tablecellbackgroundcolorcommand';
 
-import { setData as setModelData } from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
+import { getData as getModelData, setData as setModelData } from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
 import { assertEqualMarkup } from '@ckeditor/ckeditor5-utils/tests/_utils/utils';
-import { assertTableCellStyle, assertTRBLAttribute } from '../_utils/utils';
+import { assertTableCellStyle, assertTRBLAttribute, modelTable } from '../_utils/utils';
+import ModelTestEditor from '@ckeditor/ckeditor5-core/tests/_utils/modeltesteditor';
+import TableUtils from '../../src/tableutils';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tablecellproperties/tablecellpropertiesediting.js:26]: Please fix the order of imports.

File: packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js [hunk 24]
```diff
@@ -19,7 +19,9 @@ import TableBackgroundColorCommand from '../../src/tableproperties/commands/tabl
 
 import { setData as setModelData, getData as getModelData } from '@ckeditor/ckeditor5-engine/src/dev-utils/model';
 import { assertEqualMarkup } from '@ckeditor/ckeditor5-utils/tests/_utils/utils';
-import { assertTableStyle, assertTRBLAttribute } from '../_utils/utils';
+import { assertTableStyle, assertTRBLAttribute, modelTable } from '../_utils/utils';
+import ModelTestEditor from '@ckeditor/ckeditor5-core/tests/_utils/modeltesteditor';
+import TableUtils from '../../src/tableutils';
```
Comment (niegowski) [inline packages/ckeditor5-table/tests/tableproperties/tablepropertiesediting.js:24]: Please fix the order of imports.

**Review by niegowski — COMMENTED** [review]
As commented below.

**Comment by pomek** [comment]
See: https://github.com/ckeditor/ckeditor5/issues/8502#issuecomment-824222728.

**Comment by pomek** [comment]
Closing in favor of https://github.com/ckeditor/ckeditor5/pull/9555.

**Pull request closed without merging**